pkg/ccl/backupccl/restore_planning.go

```diff
@@ -158,6 +159,237 @@ func synthesizePGTempSchema(
+	for _, table := range systemTablesWithConflictingIDs {
+		log.Infof(ctx, "moving system table %s to non-conflicting ID", table.GetName())
+		if err := moveConflictingSystemTable(ctx, p, table); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// moveConflictingSystemTable moves a system table to a higher ID.
+//
+// We do the move by creating new tables in the system database. Data from the
+// conflicting tables is moved to the temporary tables. The namespace and
+// descriptor entry for the conflicting table is removed, and then the new table
+// is renamed.
+//
+// TODO(ssd): We currently do all of the descriptor modification "bare handed"
+// by writing to KV directly. Should we be using WriteDescToBatch and friends
+// instead? I avoided it for now because I was having some trouble with
+// validation when using it.
+func moveConflictingSystemTable(
+	ctx context.Context, p sql.PlanHookState, oldTable catalog.TableDescriptor,
+) error {
+	ctx, span := tracing.ChildSpan(ctx, "backupccl.moveConflictingSystemTable")
+	defer span.Finish()
+
+	codec := p.ExecCfg().Codec
+	return sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn, col *descs.Collection) error {
+		b := txn.NewBatch()
+		// Create a new table with the same schema as the table we want to
+		// move.
+		//
+		// TODO(ssd): Think about whether we need to do cleanup on failure here. I
+		// _think_ that since we do this all in a single transaction, we should be
+		// OK.
+		newTable := oldTable.NewBuilder().BuildCreatedMutable().(*tabledesc.Mutable)
+		newTable.ID = descpb.InvalidID
+		newTable.Name = fmt.Sprintf("crdb_internal_copy_%s", newTable.Name)
+		id, created, err := upgrades.CreateSystemTableInTxn(ctx, p.ExecCfg().DB, txn, codec, newTable)
+		if err != nil {
+			return err
+		}
+		if !created {
+			return errors.New("temporary table for system table already exists")
+		}
+
+		// Copy data from the old table to the new table.
+		insertStmt := fmt.Sprintf("INSERT INTO system.%s (SELECT * FROM system.%s)", newTable.Name, oldTable.GetName())
+		if _, err := p.ExecCfg().InternalExecutor.Exec(ctx, "copy-to-temp-table", txn, insertStmt); err != nil {
+			return err
+		}
+
+		// Delete the old namespace entry for the new table.
+		b.Del(catalogkeys.EncodeNameKey(codec, newTable))
+
+		// Delete the descriptor entry for the old system table.
+		b.Del(catalogkeys.MakeDescMetadataKey(codec, oldTable.GetID()))
+		col.AddDeletedDescriptor(oldTable.GetID())
+
+		// Overwrite the namespace entry for the system table with our new id, and update the
+		// descriptor of the new table with the old table's name.
+		newTable.ID = id
+		newTable.Name = oldTable.GetName()
+		b.Put(catalogkeys.EncodeNameKey(codec, newTable), newTable.ID)
+		b.Put(catalogkeys.MakeDescMetadataKey(codec, newTable.GetID()), newTable.DescriptorProto())
+
+		if err := txn.Run(ctx, b); err != nil {
+			return err
+		}
+
+		// ClearRange on the table span for the old table that we've just migrated.
```

Comment:
I added this ClearRange thinking it was prudent to clear out the keyspace we are about to potentially reuse. But the tests I added don't fail even in its absence.